systemverilog-plugin/UhdmAst.cc

```diff
@@ -2173,13 +2174,64 @@ void UhdmAst::process_typespec_member()
     add_multirange_wire(current_node, packed_ranges, unpacked_ranges);
 }
 
-void UhdmAst::process_enum_typespec()
+void UhdmAst::process_enum_typespec(const UHDM::BaseClass *object)
 {
     // BaseTypespec specifies underlying type of the enum.
     // The BaseTypespec has at most one explicit packed dimension (range).
     // When base type is not specified in SystemVerilog code, it is assumed to be an int.
     // Type of enum items (constants) is the same as the enum type.
     current_node = make_ast_node(AST::AST_ENUM);
+
+    const auto *enum_object = object->Cast<const UHDM::enum_typespec *>();
+    const auto *typespec = enum_object->Base_typespec();
+
+    if (typespec && typespec->UhdmType() == UHDM::uhdmlogic_typespec) {
+        // If it's a logic_typespec, try to reduce expressions inside of it.
+
+        const auto *logic_typespec_obj = enum_object->Base_typespec()->Cast<const UHDM::logic_typespec *>();
+        std::vector<UHDM::range *> ranges;
+        // Check if ranges exist.
+        if (logic_typespec_obj->Ranges()) {
+            ranges = logic_typespec_obj->Ranges()[0];
+        }
+        for (UHDM::range *range_obj : ranges) {
+            // For each range, take both left and right and reduce them if they're of type uhdmoperation.
+            const auto *leftrange_obj = range_obj->Left_expr();
+            const auto *rightrange_obj = range_obj->Right_expr();
+            log_assert(leftrange_obj);
+            log_assert(rightrange_obj);
+
+            if (leftrange_obj->UhdmType() == UHDM::uhdmoperation) {
+
+                const auto *operation_obj = leftrange_obj->Cast<const UHDM::operation *>();
+
+                bool invalidvalue = false;
+                UHDM::ExprEval eval;
+                UHDM::expr *resolved_operation =
+                  eval.reduceExpr(operation_obj, invalidvalue, enum_object->Instance() ? enum_object->Instance() : enum_object->VpiParent(),
+                                  enum_object->VpiParent());
+
+                log_assert(!invalidvalue);
+                // Substitute the previous leftrange with the resolved operation result.
+                range_obj->Left_expr(resolved_operation);
```

Comment:
Do we need to delete previous expression (that we are replacing here) or it is done automatically?

Reply:
I've browsed through the usages of this function in Surelog, and it appears that the old object is never deleted explicitly, and quite often it's just overwritten with the new one, so I expect this shouldn't be an issue.